File: agent/script-runner.mjs
import { pathToFileURL } from "node:url";
import { chromium, firefox } from "playwright-core";

const scriptPath = process.argv[2];
if (!scriptPath) throw new Error("script path is missing");

process.stdin.setEncoding("utf8");
process.stdin.once("end", () => process.exit(1));
process.stdin.resume();

function firstInputLine() {
  return new Promise((resolve, reject) => {
    let buffer = "";
    const onData = (chunk) => {
      buffer += chunk;
      const newline = buffer.indexOf("\n");
      if (newline === -1) return;
      process.stdin.off("data", onData);
      resolve(buffer.slice(0, newline).replace(/\r$/, ""));
    };
    const onEnd = () => {
      process.stdin.off("data", onData);
      reject(new Error("runner input is missing"));
    };
    process.stdin.on("data", onData);
    process.stdin.once("end", onEnd);
  });
}

async function run(input) {
  if (typeof input?.endpoint !== "string" || !input.endpoint) throw new Error("runner input is missing");
  if (input.endpoint.startsWith("firefox://")) throw new Error("Firefox scripts require a private Playwright endpoint");
  if (input.engine !== undefined && input.engine !== "firefox") throw new Error("runner input is invalid");
  let browser;
  try {
    browser = input.engine === "firefox"
      ? await firefox.connect(input.endpoint, { timeout: 30_000 })
      : await chromium.connectOverCDP(input.endpoint, { timeout: 30_000 });
    const context = browser.contexts()[0];
    if (!context) throw new Error("AliasMode browser context is unavailable");
    const page = context.pages()[0] ?? await context.newPage();
    const module = await import(pathToFileURL(scriptPath).href);
    if (typeof module.default !== "function") {
      throw new Error("Playwright script must export a default async function");
    }
    await module.default({
      browser,
      context,
      page,
      profile: input.profile,
      inputs: input.inputs,
      credentials: input.credentials ?? null,
      log: (...values) => console.log(...values),
    });
  } finally {
    await browser?.close().catch(() => {});
  }
}

function errorText(error, input) {
  const text = error instanceof Error ? (error.stack ?? error.message) : String(error);
  return input?.engine === "firefox" && typeof input.endpoint === "string"
    ? text.replaceAll(input.endpoint, "private Firefox endpoint")
    : text;
}

let input;
try {
  input = JSON.parse(await firstInputLine());
  await run(input);
} catch (error) {
  process.exitCode = 1;
  console.error(errorText(error, input));
} finally {
  process.stdin.pause();
}

File: agent/script-runner.py
import asyncio
import importlib.util
import inspect
import json
import os
import sys
import threading
import traceback

from playwright.async_api import async_playwright


def read_input() -> dict:
    line = sys.stdin.readline()
    if not line:
        raise RuntimeError("runner input is missing")
    value = json.loads(line)
    if not isinstance(value, dict):
        raise TypeError("runner input must be an object")
    return value


def load_script(path: str):
    spec = importlib.util.spec_from_file_location("aliasmode_user_script", path)
    if spec is None or spec.loader is None:
        raise ImportError(f"could not load script: {path}")
    module = importlib.util.module_from_spec(spec)
    sys.path.insert(0, os.path.dirname(path))
    spec.loader.exec_module(module)
    run = getattr(module, "run", None)
    if not inspect.iscoroutinefunction(run):
        raise TypeError("Python Playwright script must export async def run")
    return run


async def run_script(input_value: dict) -> None:
    browser = None
    playwright = None
    try:
        playwright = await async_playwright().start()
        if input_value.get("engine") == "firefox":
            browser = await playwright.firefox.connect(input_value["endpoint"], timeout=30_000)
        else:
            browser = await playwright.chromium.connect_over_cdp(input_value["endpoint"], timeout=30_000)
        context = browser.contexts[0] if browser.contexts else None
        if context is None:
            raise RuntimeError("AliasMode browser context is unavailable")
        page = context.pages[0] if context.pages else await context.new_page()
        run = load_script(sys.argv[1])
        await run(
            browser=browser,
            context=context,
            page=page,
            profile=input_value.get("profile"),
            inputs=input_value.get("inputs"),
            credentials=input_value.get("credentials"),
            log=lambda *values: print(*values, flush=True),
        )
    finally:
        if browser is not None:
            try:
                await browser.close()
            except Exception:
                pass
        if playwright is not None:
            try:
                await playwright.stop()
            except Exception:
                pass


def main() -> int:
    if len(sys.argv) != 2:
        raise RuntimeError("script path is missing")
    input_value = read_input()

    def watch_parent() -> None:
        sys.stdin.read()
        os._exit(1)

    threading.Thread(target=watch_parent, daemon=True).start()
    try:
        asyncio.run(run_script(input_value))
        return 0
    except BaseException:
        text = traceback.format_exc()
        if input_value.get("engine") == "firefox" and isinstance(input_value.get("endpoint"), str):
            text = text.replace(input_value["endpoint"], "private Firefox endpoint")
        sys.stderr.write(text)
        return 1


if __name__ == "__main__":
    sys.exit(main())
